Add gpu.create_stream op and add stream as an optional argument in few gpu dialect ops

nbpatel opened the pull request:
# **DO NOT FILE A PULL REQUEST**

This repository does not accept pull requests. Please follow http://llvm.org/docs/Contributing.html#how-to-submit-a-patch for contribution to LLVM.

# **DO NOT FILE A PULL REQUEST**

Hardcode84 commented on mlir/test/Dialect/GPU/ops.mlir:
```diff
@@ -234,6 +241,12 @@ module attributes {gpu.container_module} {
     // CHECK: gpu.dealloc %[[m2]] : memref<13xf32, 1>
     gpu.dealloc %m2 : memref<13xf32, 1>
 
+    %stream = gpu.create_stream : !gpu.stream
```
`// CHECK %[[stream:.*]] = gpu.create_stream : !gpu.stream`

Hardcode84 commented on mlir/test/Dialect/GPU/ops.mlir:
```diff
@@ -234,6 +241,12 @@ module attributes {gpu.container_module} {
     gpu.dealloc %m2 : memref<13xf32, 1>
 
+    %stream = gpu.create_stream : !gpu.stream
+    // CHECK: %[[m3:.*]] = gpu.alloc () stream %[[stream:.*]] : memref<13xf32, 1>
```
`%[[stream]]`

Hardcode84 commented on mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:
```diff
@@ -164,6 +166,13 @@ Type GPUDialect::parseType(DialectAsmParser &parser) const {
   if (keyword == "async.token")
     return AsyncTokenType::get(context);
 
+  if (keyword == "stream") {
```
remove curly braces

Hardcode84 commented on mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:
```diff
@@ -1327,17 +1377,22 @@ struct EraseRedundantGpuWaitOpPairs : public OpRewritePattern<WaitOp> {
     };
-    if (llvm::none_of(op.getAsyncDependencies(), predicate))
+
+    if (llvm::none_of(op.getAsyncDependencies(), predicate)) {
```
remove curly braces on single-line ifs

Commit:
[mlir][GPU] Add gpu.create_stream op and add stream as an optional argument to gpu ops

The above PR introduces a new op called gpu.create_stream and adds stream as an optional arg to some gpu dialect
ops. This will give the user the flexibility to launch kernels on their custom stream. We require this to enable
the MLIR pipeline on Intel GPU.

The link to the discussion is:
https://discourse.llvm.org/t/proposal-to-add-stream-queue-as-an-optional-argument-to-few-gpu-dialect-ops/67920/13

Differential Revision: https://reviews.llvm.org/D146556

silee2 reviewed (approved):
LGTM
